lib/engine/game/g_18_co.rb

```diff
@@ -0,0 +1,130 @@
+# frozen_string_literal: true
+
+require_relative '../config/game/g_18_co'
+require_relative 'base'
+require_relative 'company_price_50_to_150_percent'
+require_relative 'revenue_4d'
+
+module Engine
+  module Game
+    class G18CO < Base
+      register_colors(green: '#237333',
+                      red: '#d81e3e',
+                      blue: '#0189d1',
+                      lightBlue: '#a2dced',
+                      yellow: '#FFF500',
+                      orange: '#f48221',
+                      brown: '#7b352a',
+                      black: '#000000',
+                      pink: '#FF0099',
+                      purple: '#9900FF',
+                      white: '#FFFFFF')
+      load_from_json(Config::Game::G18CO::JSON)
+      AXES = { x: :number, y: :letter }.freeze
+
+      # DEV_STAGE = :beta
+
+      GAME_LOCATION = 'Colorado, USA'
+      GAME_RULES_URL = 'https://drive.google.com/open?id=0B3lRHMrbLMG_eEp4elBZZ0toYnM'
+      GAME_DESIGNER = 'R. Ryan Driskel'
+      GAME_INFO_URL = 'https://boardgamegeek.com/thread/1526930/18co-rock-stock-playtesting-files-available'
+
+      SELL_BUY_ORDER = :sell_buy
+      MUST_EMERGENCY_ISSUE_BEFORE_EBUY = true
+
+      # Two tiles can be laid, only one upgrade
+      # TODO: This changes in phase E to a single tile lay
+      TILE_LAYS = [{ lay: true, upgrade: true }, { lay: true, upgrade: false }].freeze
+
+      IPO_NAME = 'Treasury'
+
+      EVENTS_TEXT = Base::EVENTS_TEXT.merge(
+          'remove_mines' => ['Mines Close', 'Mine tokens removed from board and corporations']
+        ).freeze
+
+      include CompanyPrice50To150Percent
+      include Revenue4D
+
+      def dsng
+        @dsng ||= corporation_by_id('DSNG')
+      end
+
+      def setup
+        setup_company_price_50_to_150_percent
+
+        train = @depot.upcoming[0]
+        train.buyable = false
+        dsng.buy_train(train, :free)
+      end
+
+      def operating_round(round_num)
+        Round::Operating.new(self, [
+        Step::Bankrupt,
+        Step::DiscardTrain,
+        Step::HomeToken,
+        Step::BuyCompany,
+        Step::G18CO::Track,
+        Step::Token,
+        Step::Route,
+        Step::Dividend,
+        Step::BuyTrain,
+        [Step::BuyCompany, blocks: true],
+        ], round_num: round_num)
+      end
+
+      def stock_round
+        Round::Stock.new(self, [
+        Step::DiscardTrain,
+        Step::BuySellParShares,
+        ])
+      end
+
+      def adjust_revenue_for_5d_train(route, stops, revenue)
```

Comment:
I think it could be possible to refactor the 4D revenue thing to cover for both 4D and 5D. There are a couple of tiles that use these 4D (18AL, 18MEX, 18GA?) and then 18MS has another type of 4D.

Reply:
There is a 3+D in 1828 that is similar too. The only difference is it can also run infinite towns before, in the middle, and after the 3 cities.

Reply:
The 4D trains of 18AL/18MEX is like that. So naming them with + is a good thing. (Ie 18MEX and 18AL should probably have been called 4+D, and the normal trains in 18MEX/18AL should have been called 2+, 3+, ...)

But the naming of trains is a bit complex to make general as the "+" can have different meaning in different games.

Anyway I think it could be possible to refactor this to have reusable code for all games that use these types of Diesels, alternatively we could enhance the configuration to make it possible to specify what kind of stops are counted double? In the latter case we would not have to do any special code for that. I will submit an enhancement ticket with this suggestion.

Reply:
I fully support a refactor to allow specifying a multiplier on stop type as indicated in my original PR text. Do we wish to pursue that ahead of this PR, or loop back on it later?